---
title: "Inbred mouse SV genotyping SRP199233"
author: "Mark Ziemann & Kingsley Isaac Ogamanya"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("GenomicRanges")
library("R.utils")
library("Biostrings")
library("eulerr")
library("SummarizedExperiment")
library("VariantAnnotation")

```

## Loading the data

```{r, loaddata}

# reading in the GFF files
gff <- read.table("Mus_musculus.GRCm38.98.gff3", sep = "\t", quote = "")

v75 <- readVcf("VCF_SRP199233/SRX5884275_svtyper.vcf", "mm10")
v76 <- readVcf("VCF_SRP199233/SRX5884276_svtyper.vcf", "mm10")
v77 <- readVcf("VCF_SRP199233/SRX5884277_svtyper.vcf", "mm10")
v78 <- readVcf("VCF_SRP199233/SRX5884278_svtyper.vcf", "mm10")
v79 <- readVcf("VCF_SRP199233/SRX5884279_svtyper.vcf", "mm10")
v80 <- readVcf("VCF_SRP199233/SRX5884280_svtyper.vcf", "mm10")

r75 <- rowRanges(v75)
r76 <- rowRanges(v76)
r77 <- rowRanges(v77)
r78 <- rowRanges(v78)
r79 <- rowRanges(v79)
r80 <- rowRanges(v80)

g75 <- geno(v75)$GT
g76 <- geno(v76)$GT
g77 <- geno(v77)$GT
g78 <- geno(v78)$GT
g79 <- geno(v79)$GT
g80 <- geno(v80)$GT

```

## Reorganising the data

```{r,chunk 10}

df <- data.frame(rowRanges(v75),info(v75))  #taking a look into the content for one of the vcf file.
df <- df[,c("seqnames","start","END")]
head(df)

alt <- unlist(as.vector(elementMetadata(v75)$ALT))
svtype <- info(v75)$SVTYPE

gt <- data.frame(g75,g76,g77,g78,g79,g80)
colnames(gt) <- c("g75","g76","g77","g78","g79","g80") 

qual <- data.frame(r75$QUAL,r76$QUAL,r77$QUAL,r78$QUAL,r79$QUAL,r80$QUAL)

gt2 <- data.frame(df,alt,svtype,gt,qual) 

head(gt2,20)

```

```{r,filt1}

gt3 <- gt2[which(apply(qual,1,min)>=20),]
head(gt3,20)

gt <- gt[which(apply(qual,1,min)>=20),]
head(gt,20)

rd <- rowRanges(v75)
rd <- rd[which(apply(qual,1,min)>=20),]
head(rd,20)

rows <- which(apply(gt,1,function(x){
  sd(as.numeric(as.factor(x)))
})!=0)

gt3 <- gt3[rows,]
gt <- gt[rows,]

gtx <- apply(gt,2,function(x) { as.numeric(as.factor(x))})
rownames(gtx) <- paste(gt3$seqnames,gt3$start,gt3$END,gt3$alt)
gtx
myrownames <- rownames(gtx)
myregions <- do.call(rbind,strsplit(myrownames," "))
write.table(x=myregions,file="myregions.bed",quote=FALSE,sep="\t", col.names = FALSE, row.names = FALSE)

rd <- rd[rows,]
rd

```


```{r,heat1}
colfunc <- colorRampPalette(c("blue", "red"))
library(gplots)
heatmap.2(gtx, trace = "none", scale = "none", 
  dendrogram='none', Rowv=FALSE, Colv=FALSE, col=colfunc(25),
  margins = c(5,20), cexRow=.8, cexCol=.9,  main="Genetic Variation")

```


```{r,mycircos,fig.height=7,fig.width=7}

dels <- subset(gt3,svtype=="DEL")[,1:3]
dels$seqnames <- gsub("^","chr",dels$seqnames)
dups <- subset(gt3,svtype=="DUP")[,1:3]
dups$seqnames <- gsub("^","chr",dups$seqnames)
brks <- subset(gt3,svtype=="BND")[,c(1,2,2)]
brks <- brks[grep("GL",brks$seqnames,invert = TRUE),]
brks$seqnames <- gsub("^","chr",brks$seqnames)
brks

links <- subset(gt3,svtype=="BND")
links <- links[grep("GL",links$seqnames,invert = TRUE),]
links$alt <- gsub("\\[","\\@",links$alt)
links$alt <- gsub("\\]","\\@",links$alt)
links$alt <- sapply(strsplit(links$alt,"@"),"[[",2)
links$chr2 <- sapply(strsplit(links$alt,":"),"[[",1)
links$start2 <- as.numeric(sapply(strsplit(links$alt,":"),"[[",2))
links <- links[,c("seqnames","start","start","chr2","start2","start2")]
links$seqnames <- gsub("^","chr",links$seqnames)
links$chr2 <- gsub("^","chr",links$chr2)
links


library("RCircos")
data(UCSC.Mouse.GRCm38.CytoBandIdeogram)
head(UCSC.Mouse.GRCm38.CytoBandIdeogram)

RCircos.Set.Core.Components(
	cyto.info=UCSC.Mouse.GRCm38.CytoBandIdeogram,
	tracks.inside=4, tracks.outside=0)

RCircos.Set.Plot.Area()
RCircos.Chromosome.Ideogram.Plot()

RCircos.Tile.Plot(tile.data=dels, track.num=1, side="in")

RCircos.Tile.Plot(tile.data=dups, track.num=2, side="in")

RCircos.Tile.Plot(tile.data=links, track.num=3, side="in")

RCircos.Link.Plot(link.data=links, track.num = 4)

```



```{r, locate variants}

#using the package below to locate and annotate the gene variants without necessary needing the gff file at the moment
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene

```


```{r,loadgff}

gff_file <- "Mus_musculus.GRCm38.98.gff3"
txdb <- makeTxDbFromGFF(gff_file, format="gff3")

```

## Transcript intercept

```{r,txint}

tx <- transcripts(txdb)
ol <- findOverlaps(rd,tx)
ol
var_ol <- rd[queryHits(ol)]
tx_ol <- tx[subjectHits(ol)]
var_ol
tx_ol
vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol,use.mcols = TRUE)))) 
head(vardf)
txdf <- as.data.frame(tx_ol,row.names = 1:nrow(as.data.frame(ranges(tx_ol))))
head(txdf)
vardf$tx <- txdf$tx_name
head(vardf)
vardf_tx <- vardf
head(vardf_tx)
dim(vardf_tx)


```

## Gene intercept

```{r,gene int}

gx <- genes(txdb)
ol <- findOverlaps(rd,gx,select = "all")
ol

var_ol <- rd[queryHits(ol)]
gx_ol <- gx[subjectHits(ol)]
var_ol

vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol))))
vardf$ALT <- as.character(GenomicRanges::elementMetadata(var_ol)$ALT)
gxdf <- as.data.frame(gx_ol,row.names = 1:nrow(as.data.frame(ranges(gx_ol))))
vardf$gene <- gxdf$gene_id
vardf
dim(vardf)
write.table(x=vardf,file="vardf.tsv",quote = FALSE,sep = "\t",row.names = FALSE)

```


## Session Information

```{r, session}

sessionInfo()

```
