---
title: "Inbred mouse SV genotyping SRP199233"
author: "Mark Ziemann & Kingsley Isaac Ogamanya"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("GenomicRanges")
library("R.utils")
library("Biostrings")
library("eulerr")
library("SummarizedExperiment")
library("VariantAnnotation")

```

## Loading the data

```{r, loaddata}

# reading in the GFF files
gff <- read.table("Mus_musculus.GRCm38.98.gff3", sep = "\t", quote = "")

v75 <- readVcf("VCF_SRP199233_GRCm39/SRX5884275_svtyper.vcf", "GRCm39")
v76 <- readVcf("VCF_SRP199233_GRCm39/SRX5884276_svtyper.vcf", "GRCm39")
v77 <- readVcf("VCF_SRP199233_GRCm39/SRX5884277_svtyper.vcf", "GRCm39")
v78 <- readVcf("VCF_SRP199233_GRCm39/SRX5884278_svtyper.vcf", "GRCm39")
v79 <- readVcf("VCF_SRP199233_GRCm39/SRX5884279_svtyper.vcf", "GRCm39")
v80 <- readVcf("VCF_SRP199233_GRCm39/SRX5884280_svtyper.vcf", "GRCm39")

r75 <- rowRanges(v75)
r76 <- rowRanges(v76)
r77 <- rowRanges(v77)
r78 <- rowRanges(v78)
r79 <- rowRanges(v79)
r80 <- rowRanges(v80)

g75 <- geno(v75)$GT
g76 <- geno(v76)$GT
g77 <- geno(v77)$GT
g78 <- geno(v78)$GT
g79 <- geno(v79)$GT
g80 <- geno(v80)$GT

```

## Reorganising the data

```{r,chunk 10}

df <- data.frame(rowRanges(v75),info(v75))  #taking a look into the content for one of the vcf file.
df <- df[,c("seqnames","start","END")]
head(df)

alt <- unlist(as.vector(elementMetadata(v75)$ALT))
svtype <- info(v75)$SVTYPE

gt <- data.frame(g75,g76,g77,g78,g79,g80)
colnames(gt) <- c("g75","g76","g77","g78","g79","g80") 

qual <- data.frame(r75$QUAL,r76$QUAL,r77$QUAL,r78$QUAL,r79$QUAL,r80$QUAL)

gt2 <- data.frame(df,alt,svtype,gt,qual) 

head(gt2,20)

```

```{r,filt1}

gt3 <- gt2[which(apply(qual,1,min)>=20),]
head(gt3,20)

gt <- gt[which(apply(qual,1,min)>=20),]
head(gt,20)

rd <- rowRanges(v75)
rd <- rd[which(apply(qual,1,min)>=20),]
head(rd,20)

rows <- which(apply(gt,1,function(x){
  sd(as.numeric(as.factor(x)))
})!=0)

gt3 <- gt3[rows,]
gt <- gt[rows,]

gtx <- apply(gt,2,function(x) { as.numeric(as.factor(x))})
rownames(gtx) <- paste(gt3$seqnames,gt3$start,gt3$END,gt3$alt)
gtx

rd <- rd[rows,]
rd

```


```{r,heat1}
colfunc <- colorRampPalette(c("blue", "red"))
library(gplots)
heatmap.2(gtx, trace = "none", scale = "none", 
  dendrogram='none', Rowv=FALSE, Colv=FALSE, col=colfunc(25),
  margins = c(5,20), cexRow=.8, cexCol=.9,  main="Genetic Variation")

```


## Session Information

```{r, session}

sessionInfo()

```
