---
title: "Research studies on SRP_199233 using the mm39 reference genome"
author: "Kingsley Isaac Ogamanya & Mark Ziemann"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    fig_width: 7
    fig_height: 7
theme: cosmo
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("GenomicRanges")
library("R.utils")
library("Biostrings")
library("eulerr")
library("SummarizedExperiment")
library("VariantAnnotation")

```

# Loading the 6 samples and also the GFF files

```{r, loadata}
# reading in the GFF files
#download.file("http://ftp.ensembl.org/pub/release-104/gff3/mus_musculus/Mus_musculus.GRCm39.104.gff3.gz",destfile = "Mus_musculus.GRCm39.104.gff3.gz")

#library("R.utils")

#gunzip("Mus_musculus.GRCm39.104.gff3.gz")

gff <- read.table("Mus_musculus.GRCm39.104.gff3", sep = "\t", quote = "")

#Reading in the 6animals using the most investigated suitable package
v275 <- readVcf("VCF_SRP199233_GRCm39/SRX5884275_svtyper.vcf", "GRCm39")
v276 <- readVcf("VCF_SRP199233_GRCm39/SRX5884276_svtyper.vcf", "GRCm39")
v277 <- readVcf("VCF_SRP199233_GRCm39/SRX5884277_svtyper.vcf", "GRCm39")
v278 <- readVcf("VCF_SRP199233_GRCm39/SRX5884278_svtyper.vcf", "GRCm39")
v279 <- readVcf("VCF_SRP199233_GRCm39/SRX5884279_svtyper.vcf", "GRCm39")
v280 <- readVcf("VCF_SRP199233_GRCm39/SRX5884280_svtyper.vcf", "GRCm39")

#Having a look at the content of the 6 animals using the `rowRanges` function
r275 <- rowRanges(v275)
r276 <- rowRanges(v276)
r277 <- rowRanges(v277)
r278 <- rowRanges(v278)
r279 <- rowRanges(v279)
r280 <- rowRanges(v280)

#Analyzing the geno of the 6 animals 
g275 <- geno(v275)$GT
g276 <- geno(v276)$GT
g277 <- geno(v277)$GT
g278 <- geno(v278)$GT
g279 <- geno(v279)$GT
g280 <- geno(v280)$GT

```

# Reorganising and subsetting the data 

```{r, chunk 2}
df <- data.frame(rowRanges(v275),info(v275))  #taking a look into the content for one of the vcf file.
df <- df[,c("seqnames","start","END")]
head(df)

alt <- unlist(as.vector(elementMetadata(v275)$ALT))
svtype <- info(v275)$SVTYPE

gt <- data.frame(g275,g276,g277,g278,g279,g280)
colnames(gt) <- c("g275","g276","g277","g278","g279","g280") 
head(gt)

qual <- data.frame(r275$QUAL,r276$QUAL,r277$QUAL,r278$QUAL,r279$QUAL,r280$QUAL)

gt2 <- data.frame(df,alt,svtype,gt,qual)
#gt2
head(gt2,20)

#plotting a venn diagram showing only homozygous variants
a_set <- which(g275 == "1/1")
b_set <- which(g276 == "1/1")
x_set <- which(g277 == "1/1")
y_set <- which(g278 == "1/1")
d_set <- which(g279 == "1/1")
e_set <- which(g280 == "1/1")

v1 <- list("a" = a_set, "b" = b_set, "x" = x_set, "y" = y_set, "d" = d_set, "e" = e_set)
print(plot(euler(v1),quantities = TRUE,main="homozygous variants"))

```

# Creating a function that can analyse two samples

```{r, function,results='hide'}
# A function comparing two animals together with a visualization on venn diagram for possible future reference.
compare_vcf <- function(vcf_a,vcf_b) {
    gta <- geno(vcf_a)$GT
    gtb <- geno(vcf_b)$GT
    df <- data.frame(rowRanges(vcf_a),info(vcf_a))
    df <- df[,c("seqnames","start","END","SVTYPE","SVLEN","QUAL")]
    gt <- data.frame(gta,gtb)
    colnames(gt) <- c("a","b")
    gt <- data.frame(df,gt)
    nr <- nrow(gt)
    gt <- subset(gt, a != "./." & b != "./.")
    gtnr <- nrow(gt)
    missing <- nr - gtnr
    gtsame <- length(which(gt$a == gt$b))
    gtdiff <- length(which(gt$a != gt$b))
    barplot(table(gt$a),main="a")
    barplot(table(gt$b),main="b")
    gt2 <- subset(gt, a != "0/1" & b != "0/1")
    aset <- which(gt2$a == "1/1")
    bset <- which(gt2$b == "1/1")
    v1 <- list("a" = aset , "b" = bset)
    print(plot(euler(v1),quantities = TRUE,main="homozygous variants"))
    output=list("gt"=gt, "missing"=missing, "gtsame"=gtsame,"gtdiff"=gtdiff)
    return(output)
}

#testing the new function if it works perfectly.
res <- compare_vcf(v275,v276)

str(res)
head(res)
```

# proceding with analyzing all 6 Animals together 

```{r, filt1}
#Performing filtering on the 6 samples
gt3 <- gt2[which(apply(qual,1,min)>=20),]
head(gt3,20)

gt <- gt[which(apply(qual,1,min)>=20),]
head(gt,20)

rd <- rowRanges(v275)
rd <- rd[which(apply(qual,1,min)>=20),]
head(rd,20)

rows <- which(apply(gt,1,function(x){
  sd(as.numeric(as.factor(x)))
})!=0)

gt3 <- gt3[rows,]
gt <- gt[rows,]

gtx <- apply(gt,2,function(x) { as.numeric(as.factor(x))})
rownames(gtx) <- paste(gt3$seqnames,gt3$start,gt3$END,gt3$alt)
gtx
#extracting the bed files for the variants regions of this study for further investigation on IGV
myrownames <- rownames(gtx)
srp_regions <- do.call(rbind,strsplit(myrownames," "))
write.table(x=srp_regions,file="Srp_regions.bed",quote=FALSE,sep="\t", col.names = FALSE,row.names = FALSE)

rd <- rd[rows,]
rd

#from my investigation so far, it is evident that variants with high confidence can only be identify through there quality check
```


```{r,heat1, fig.width=8, fig.height=8}
# Plotting a heatmap
colfunc <- colorRampPalette(c("blue", "red"))
library(gplots)
heatmap.2(gtx, trace = "none", scale = "none", 
  dendrogram='none', Rowv=FALSE, Colv=FALSE, col=colfunc(25),
  margins = c(5,20), cexRow=.8, cexCol=.9,  main="Genetic Variation")

# when subset to 30 or higher numbers, less variants shown on the heat-map
```


```{r, locate variants}
#using the package below to locate and annotate the gene variants without necessary needing the gff file at the moment
library(TxDb.Mmusculus.UCSC.mm39.refGene) 

```

```{r,loadgff, results='hide'}

gff_file <- "Mus_musculus.GRCm39.104.gff3"
txdb <- makeTxDbFromGFF(gff_file, format="gff3")

```

# Transcript Intercept

```{r,txint, results='hide'}

tx <- transcripts(txdb)
ol <- findOverlaps(rd,tx)
ol
var_ol <- rd[queryHits(ol)]
tx_ol <- tx[subjectHits(ol)]
var_ol
tx_ol
vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol,use.mcols = TRUE)))) 
head(vardf)
txdf <- as.data.frame(tx_ol,row.names = 1:nrow(as.data.frame(ranges(tx_ol))))
head(txdf)
vardf$tx <- txdf$tx_name
head(vardf)
vardf_tx <- vardf
head(vardf_tx)
dim(vardf_tx)

```

# Gene Intercept

```{r,gene int, results='hide'}

gx <- genes(txdb)
ol <- findOverlaps(rd,gx,select = "all")
ol

var_ol <- rd[queryHits(ol)]
gx_ol <- gx[subjectHits(ol)]
var_ol

vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol))))
vardf$ALT <- as.character(GenomicRanges::elementMetadata(var_ol)$ALT)
head(vardf)

gxdf <- as.data.frame(gx_ol,row.names = 1:nrow(as.data.frame(ranges(gx_ol))))
head(gxdf)
vardf$gene <- gxdf$gene_id

head(vardf)
write.table(x=vardf,file="vardf_srp.tsv",quote = FALSE,sep = "\t",row.names = FALSE)

```

# Locating and merging the gene names 

```{r, gene names}

genenames <- read.table("Mus_musculus.GRCm39.104.gff3.genenames.tsv", header = FALSE, sep = "\t", quote = "")
colnames(genenames) <- c("Ensembl ID","Gene Name")
head(genenames)

gene_variants <- merge(genenames, vardf,by.x="Ensembl ID", by.y="gene")
head(gene_variants)
gene_variants

#At the start of the investigation it is evident that all samples has 4 types of variation, which are Deletions, Inversion, Duplication and Break-ends. 
del_variants <- subset(gene_variants, ALT=="<DEL>")
del_variants$seqnames <- gsub("^","chr",del_variants$seqnames)
del_variants

dup_variants <- subset(gene_variants, ALT=="<DUP>")
dup_variants$seqnames <- gsub("^","chr",dup_variants$seqnames)
dup_variants

inv_variants <- subset(gene_variants, ALT=="<INV>")
inv_variants$seqnames <- gsub("^","chr",inv_variants$seqnames)
inv_variants

breaks_variants <- subset(gene_variants, ALT!="<DUP>" & ALT!="<DEL>" & ALT!="<INV>")
breaks_variants$seqnames <- gsub("^","chr",breaks_variants$seqnames)
breaks_variants <- breaks_variants[order(-breaks_variants$QUAL),]
breaks_variants
```

# Circos Plot

```{r,mycircos,fig.height=7,fig.width=7}

#Subsetting for a circos plot, showing all variations except inversion
del_variants <- subset(gt3,svtype=="DEL")[,1:3]
del_variants$seqnames <- gsub("^","chr",del_variants$seqnames)
dup_variants <- subset(gt3,svtype=="DUP")[,1:3]
dup_variants$seqnames <- gsub("^","chr",dup_variants$seqnames)
breaks_variants <- subset(gt3,svtype=="BND")[,c(1,2,2)]
breaks_variants <- breaks_variants[grep("GL",breaks_variants$seqnames,invert = TRUE),]
breaks_variants$seqnames <- gsub("^","chr",breaks_variants$seqnames)
breaks_variants

links <- subset(gt3,svtype=="BND")
links <- links[grep("GL",links$seqnames,invert = TRUE),]
links$alt <- gsub("\\[","\\@",links$alt)
links$alt <- gsub("\\]","\\@",links$alt)
links$alt <- sapply(strsplit(links$alt,"@"),"[[",2)
links$chr2 <- sapply(strsplit(links$alt,":"),"[[",1)
links$start2 <- as.numeric(sapply(strsplit(links$alt,":"),"[[",2))
links <- links[,c("seqnames","start","start","chr2","start2","start2")]
links$seqnames <- gsub("^","chr",links$seqnames)
links$chr2 <- gsub("^","chr",links$chr2)
links

library("RCircos")
# performing a mouse genome liftover for the Rcircos plot
nbedfile <- read.table("SRP_hglft_genome.bed", sep = "\t", quote = "")
#nbedfile <- subset(nbedfile, V1 != "chrUn_GL456393" & V1 != "chr4_GL456216_random")
nbedfile$V1 <- factor(nbedfile$V1)
#which(nbedfile$V1 %in% unique(UCSC.Mouse.GRCm38.CytoBandIdeogram$V1))
head(nbedfile)
data(UCSC.Mouse.GRCm38.CytoBandIdeogram) #decision and answers needed for the Circos mouse genome
head(UCSC.Mouse.GRCm38.CytoBandIdeogram)

RCircos.Set.Core.Components(
	cyto.info=UCSC.Mouse.GRCm38.CytoBandIdeogram,
	tracks.inside=5, tracks.outside=0)

RCircos.Set.Plot.Area()
RCircos.Chromosome.Ideogram.Plot()

RCircos.Tile.Plot(tile.data=nbedfile, track.num=1, side="in")

RCircos.Tile.Plot(tile.data=del_variants, track.num=2, side="in")

RCircos.Tile.Plot(tile.data=dup_variants, track.num=3, side="in")

RCircos.Tile.Plot(tile.data=breaks_variants, track.num=4, side="in")

RCircos.Link.Plot(link.data=links, track.num = 5)

```

# Session Information

```{r, session}

sessionInfo()

```
