---
title: "Working on project studies"
author: "Kingsley/Mark"
date: "July 17, 2021"
output:
  html_document: default
  rmarkdown::html_vignette: default
  pdf_document: default
package: VariantAnnotation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library("GenomicRanges")
library("R.utils")
library("Biostrings")
library("eulerr")
library("SummarizedExperiment")
library("VariantAnnotation")

```

# Using the VariantAnnotation package in accordance with the package workflow

## Processing the first two vcf file of SRP199233

```{r, chunk 1}
#reading in the first two VCF file with VariantAnnotation

vcf_x <- readVcf("VCF_SRP199233/SRX5884275_svtyper.vcf", "mm10")

vcf_y <- readVcf("VCF_SRP199233/SRX5884276_svtyper.vcf", "mm10")
```

```{r, chunk 2}
#downloading the gff files and installing a package to unzip the gff file.

#download.file("ftp://ftp.ensembl.org/pub/release-98/gff3/mus_musculus/Mus_musculus.GRCm38.98.gff3.gz",destfile = "Mus_musculus.GRCm38.98.gff3.gz")

#library("R.utils")

#gunzip("Mus_musculus.GRCm38.98.gff3.gz")

```

```{r, chunk 3}
# reading in the GFF files
gff <- read.table("Mus_musculus.GRCm38.98.gff3", sep = "\t", quote = "")
```

```{r, chunk 4}

# this is a better way to look at the contents of the VCF files
rowRanges(vcf_x)
rd_x <- rowRanges(vcf_x)
info(vcf_x)
info(VariantAnnotation::header(vcf_x))
geno(vcf_x)
geno(VariantAnnotation::header(vcf_x))
header(vcf_x)
samples(VariantAnnotation::header(vcf_x))


#determining the genomic positions of the data using the rowRanges function which contains information from the CHROM, POS and ID fields of the Vcf files.
rowRanges(vcf_x)
head(rowRanges(vcf_x, 3))

head(rowRanges(vcf_y, 3))
rd_y <- rowRanges(vcf_y)
```

```{r, chunk 5}
#looking into the individuals fields of both vcf file for the first length 9
ref(vcf_x)[1:9]
qual(vcf_x)[1:9]
alt(vcf_x)[1:9]

ref(vcf_y)[1:9]
qual(vcf_y)[1:9]
alt(vcf_y)[1:9]
```

```{r, chunk 6, results='hide'}
#getting the full list of genotype in both vcf files and describing them in `FORMAT` fields as dataframe and class
geno(vcf_x)
sapply(geno(vcf_x), DataFrame)
sapply(geno(vcf_x), class)


geno(vcf_y)
sapply(geno(vcf_y), DataFrame)
sapply(geno(vcf_y), class)

#I observed lots of differences with the genotype of both vcf when described in dataframe formats
```

```{r, chunk 7, results='hide'}
#having a look at the genotype (GT) by analysing the data as a factor 
geno(VariantAnnotation::header(vcf_x))["GT",]
GT_x <- geno(vcf_x)$GT
GT_x <- as.factor(GT_x)
head(GT_x)
str(GT_x)
table(GT_x)


geno(VariantAnnotation::header(vcf_y))["GT",]
GT_y <- geno(vcf_y)$GT
GT_y <- as.factor(GT_y)
head(GT_y)
str(GT_y)
table(GT_y)
```

```{r, chunk 8, results='hide'}
#comparing the two animals genotype as a single data frame 
GT <- data.frame(GT_x, GT_y)
head(GT)
GT[,c("GT_x","GT_y")]
str(GT)

#filtering the missing data found within the variants
subset(GT, GT_x != "./." & GT_y != "./.")
GT_f <- subset(GT, GT_x != "./." & GT_y != "./.")
GT_f

#using table and barplot to check the level of heterozygous and homozygous found within the two animals
table(paste(GT_f$GT_x,GT_f$GT_y))
barplot(table(GT_f$GT_x == GT_f$GT_y))
```


```{r, chunk 9}
#creating a function that compares two animals together.
compare_vcf <- function(vcf_a,vcf_b) {
    gta <- geno(vcf_a)$GT
    gtb <- geno(vcf_b)$GT
    df <- data.frame(rowRanges(vcf_a),info(vcf_a))
    df <- df[,c("seqnames","start","END","SVTYPE","SVLEN","QUAL")]
    gt <- data.frame(gta,gtb)
    colnames(gt) <- c("a","b")
    gt <- data.frame(df,gt)
    nr <- nrow(gt)
    gt <- subset(gt, a != "./." & b != "./.")
    gtnr <- nrow(gt)
    missing <- nr - gtnr
    gtsame <- length(which(gt$a == gt$b))
    gtdiff <- length(which(gt$a != gt$b))
    barplot(table(gt$a),main="a")
    barplot(table(gt$b),main="b")
    gt2 <- subset(gt, a != "0/1" & b != "0/1")
    aset <- which(gt2$a == "1/1")
    bset <- which(gt2$b == "1/1")
    v1 <- list("a" = aset , "b" = bset)
    print(plot(euler(v1),quantities = TRUE,main="homozygous variants"))
    output=list("gt"=gt, "missing"=missing, "gtsame"=gtsame,"gtdiff"=gtdiff)
    return(output)
}

#function created and to be tried on two different animals read with VariantAnnotation package.
vcf_a <- readVcf("VCF_SRP199233/SRX5884277_svtyper.vcf", "mm10")
rd_a <- rowRanges(vcf_a)

vcf_b <- readVcf("VCF_SRP199233/SRX5884278_svtyper.vcf", "mm10")
rd_b <- rowRanges(vcf_b)

#attempting the new function.
res <- compare_vcf(vcf_a,vcf_b)

#str(res)
#head(res)
```

Analyzing all 6 Animals together 
```{r,chunk 10}
#reading in the last 2 vcf data.
vcf_d <- readVcf("VCF_SRP199233/SRX5884279_svtyper.vcf", "mm10")
rd_d <- rowRanges(vcf_d)

vcf_e <- readVcf("VCF_SRP199233/SRX5884280_svtyper.vcf", "mm10")
rd_e <- rowRanges(vcf_e)

#analyzing the geno for the 6 animals respectively.

gtx <- geno(vcf_x)$GT
gty <- geno(vcf_y)$GT
gta <- geno(vcf_a)$GT
gtb <- geno(vcf_b)$GT
gtd <- geno(vcf_d)$GT
gte <- geno(vcf_e)$GT

df <- data.frame(rowRanges(vcf_x),info(vcf_x))  #taking a look into the content for one of the vcf file.
head(df)
df <- df[,c("seqnames","start","END","SVTYPE","SVLEN","QUAL")]
gt <- data.frame(gtx,gty,gta,gtb,gtd,gte)
gt <- data.frame(df,gt) # this gives us the quality check of each variants for the 6 animals and from here i will need to copy objects gt into gt2 and then subset gt2 for the variants with highest confidence using the QUAL which should be > 30.
gt2 <- gt

colnames(gt) <- c("x","y","a","b","d","e") #continue working on gt objects to get the true Homozygous variants for all animals and visualize it on a venn diagram
gt <- subset(gt != "./." & gt != "0/1")
head(gt)
table(gt) #having a tabular look of the TRUE Homozygous variants for all 6 animals.

a_set <- which(gta == "1/1")
b_set <- which(gtb == "1/1")
x_set <- which(gtx == "1/1")
y_set <- which(gty == "1/1")
d_set <- which(gtd == "1/1")
e_set <- which(gte == "1/1")

v1 <- list("a" = a_set, "b" = b_set, "x" = x_set, "y" = y_set, "d" = d_set, "e" = e_set)
print(plot(euler(v1),quantities = TRUE,main="homozygous variants"))


```

```{r, heatmap1, fig.width=8, fig.height=8}
gt2 <- subset(gt2, gt2$QUAL>30)

gt3 <- apply(gt2[,c(7:12)], 2, function(x){
  as.numeric(as.factor(x))-2
})
cbind(gt2,gt3)

gt2[gt2 == -1] <- NA

library(gplots)

rownames(gt3) <- paste(gt2$seqnames,gt2$start,sep=":")
rownames(gt3)
colfunc <- colorRampPalette(c("yellow", "blue", "red"))
heatmap.2(gt3, trace = "none", scale = "none", dendrogram='none', Rowv=FALSE, Colv=FALSE, col=colfunc(25), margins = c(5,10), cexRow=.8, cexCol=.9,  main="Genetic Variation")

#from my little investigation, i could tell that variants with high confidence can only be identify through there quality check
```

```{r, locate variants}
#using the package below to locate and annotate the gene variants without necessary needing the gff file at the moment
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
#txdb <- TxDb.Mmusculus.UCSC.mm10.ensGene


#myseqlevels <- gsub("chrUn_","",seqlevels(txdb))
#myseqlevels <- gsub("_random","",myseqlevels)
#strReverse <- function(x) { sapply(lapply(strsplit(x, NULL), rev), paste, collapse="") }
#myseqlevels <- strReverse(myseqlevels)
#myseqlevels <- sapply(strsplit(myseqlevels , "_"),"[[",1)
#myseqlevels <- strReverse(myseqlevels)
#myseqlevels <- gsub("chr","",myseqlevels)
#myseqlevels <- gsub("M","MT",myseqlevels)
#myseqlevels


# why are the variants in rd only 1bp size?
#loc <- locateVariants(rd, txdb, CodingVariants())

#all <- locateVariants(rd, txdb, AllVariants())

```


```{r,loadgff}

gff_file <- "Mus_musculus.GRCm38.98.gff3"
txdb <- makeTxDbFromGFF(gff_file, format="gff3")

vcfd <- vcf_d #copied vcf_d file to vcfd, so as to use vcfd to have a clear look into how to annotate and locate the gene variants without altering the vcf_d files.

rd <- rowRanges(vcfd)

rd
rd <- rd[elementMetadata(rd)[,4]>30]


txdb
seqlevels(vcfd)
seqlevels(txdb)

intersect(seqlevels(rd),seqlevels(txdb))

```

## transcript intercept

```{r,txint}

tx <- transcripts(txdb)
ol <- findOverlaps(rd,tx)
ol
var_ol <- rd[queryHits(ol)]
tx_ol <- tx[subjectHits(ol)]
var_ol
tx_ol
vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol,use.mcols = TRUE)))) #read up more on the tab options to get the ALT.
head(vardf)
txdf <- as.data.frame(tx_ol,row.names = 1:nrow(as.data.frame(ranges(tx_ol))))
head(txdf)
vardf$tx <- txdf$tx_name
head(vardf)
vardf_tx <- vardf
vardf_tx
```

## gene intercept

```{r,gene int}

gx <- genes(txdb)
ol <- findOverlaps(rd,gx)
ol
var_ol <- rd[queryHits(ol)]
gx_ol <- gx[subjectHits(ol)]
var_ol
gx_ol
vardf <- as.data.frame(var_ol,row.names = 1:nrow(as.data.frame(ranges(var_ol))))
head(vardf)
gxdf <- as.data.frame(gx_ol,row.names = 1:nrow(as.data.frame(ranges(gx_ol))))
head(gxdf)
vardf$gene <- gxdf$gene_id
vardf

```

#locating and merging the gene names 
```{r, gene names}

genenames <- read.table("Mus_musculus.GRCm38.98.gff3.genenames.tsv", header = FALSE, sep = "\t", quote = "")
colnames(genenames) <- c("Ensembl ID","Gene Name")
genenames

gene_variants <- merge(genenames, vardf,by.x="Ensembl ID", by.y="gene")
gene_variants

#subset from the gene variants objects to get the 3 types of variation to show on the circos plot

library("RCircos")


```

```{r, session}
sessionInfo()
```



