---
title: "using the vcf package"
author: "Kingsley"
date: "17/03/2021"
output: html_document
---

```{r,setup}
pkg <- "pinfsc50"
vcf_file <- system.file("extdata", "pinf_sc50.vcf.gz", package = pkg)
dna_file <- system.file("extdata", "pinf_sc50.fasta", package = pkg)
gff_file <- system.file("extdata", "pinf_sc50.gff", package = pkg)
knitr::opts_chunk$set(echo = TRUE)
library(vcfR)
vcf <- read.vcfR( vcf_file, verbose = FALSE )
```

## Introduction

This is a VCF package meant to execute codes in Rmarkdown

```{r cars}
dna <- ape::read.dna(dna_file, format = "fasta")
gff <- read.table(gff_file, sep="\t", quote="")
```

## Including Plots

```{r, plot1}
library(vcfR)
chrom <- create.chromR(name='Supercontig', vcf=vcf, seq=dna, ann=gff)
plot(chrom)
```

```{r, plot2}
chrom <- masker(chrom, min_QUAL = 1, min_DP = 300, max_DP = 700, min_MQ = 59.9,  max_MQ = 60.1)
plot(chrom)
```

```{r, plot3}
chrom <- proc.chromR(chrom, verbose=TRUE)
plot(chrom)
```

```{r,chrom1}
chromoqc(chrom, dp.alpha=20)
```

```{r,chrom2}
chromoqc(chrom, xlim=c(5e+05, 6e+05))
```

session Information

```{r pressure, echo=FALSE}
sessionInfo()
```

